### DISCUSSION TITLE
Can json payloads be strongly typed? Or should all the fields be String?

### DISCUSSION THREAD

**Discussion by matthiasdebernardini** [tokio-rs/axum#2130] [Q&A] [answered]
Lets say I have a route and I want to accept a Json payload like so;

```rust
pub async fn check_sigs(Json(v): Json<SpecialType>) -> Result<Json<Value>, AppError> {}
```

`SpecialType` is a custom type, like so;

```rust
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SpecialType {
    pub context: Context,
    pub siglist: Vec<Siglist>,
}
```

Is this possible? I can't find an example that has something like it.

The purpose of this is to avoid having to cast the type in my handler and move that code out elsewhere.

Currently, I can only get it to work like this, where instead of my custom type, I have `String`. `Siglist` is a type made of integers and strings, so it just works. My `Context` type had more custom types.

```rust
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SpecialType {
    pub context: String,
    pub siglist: Vec<Siglist>,
}
```

**Comment by davidpdrsn** [comment]
Yea you can use custom types in JSON, that’s just using serde.

JSON only supports strings as map keys but otherwise it should work.

**Reply by matthiasdebernardini** [reply]
Thank you for your haste in answering, do you have an example of using custom types with Json in the payload for axum?

**Reply by davidpdrsn** [reply]
I don't understand what issue you're having. Probably an https://xyproblem.info

```rust
#[derive(Serialize)]
struct Foo {
    bar: Bar,
}

#[derive(Serialize)]
struct Bar {
    n: i32,
}

async fn get_foo() -> Json<Foo> {
    Json(Foo {
        bar: Bar {
            n: 123,
        }
    })
}
```

**Reply by matthiasdebernardini** [reply]
Indeed it can (I just tried to recreate a minimal example and realized this), the issue lies with the types that I am trying to use not axum. They are base64 encoded strings which is causing the problem.

**Comment by matthiasdebernardini** [comment, marked as answer]
Ok for posterity, my issue is that the underlying crate expands the the type in an unexpected way.

The solution to this is to annotate the field in the struct with the problematic type with `#[serde(deserialize_with = "my_deserialize_fn")]`
